bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java

```diff
@@ -0,0 +1,322 @@
+import org.apache.bookkeeper.proto.checksum.DigestManager;
+import org.apache.bookkeeper.util.ByteBufList;
+import org.apache.bookkeeper.util.MathUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BatchedReadOp extends ReadOpBase implements BatchedReadEntryCallback {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BatchedReadOp.class);
+
+    final int maxCount;
+    final long maxSize;
+
+    BatchedLedgerEntryRequest request;
+
+    BatchedReadOp(LedgerHandle lh,
+                  ClientContext clientCtx,
+                  long startEntryId,
+                  int maxCount,
+                  long maxSize,
+                  boolean isRecoveryRead) {
+        super(lh, clientCtx, startEntryId, -1L, isRecoveryRead);
+        this.maxCount = maxCount;
+        this.maxSize = maxSize;
+    }
+
+    @Override
+    void initiate() {
+        this.requestTimeNanos = MathUtils.nowInNano();
+        List<BookieId> ensemble = getLedgerMetadata().getEnsembleAt(startEntryId);
+        if (parallelRead) {
+            LOG.info("Batch read unsupported the parallelRead, failback to sequence read.");
+        }
+        request = new SequenceReadRequest(ensemble, lh.ledgerId, startEntryId, maxCount, maxSize);
+        request.read();
+    }
+
+    @Override
+    protected void submitCallback(int code) {
+        // ensure callback once
+        if (!complete.compareAndSet(false, true)) {
+            return;
+        }
+
+        long latencyNanos = MathUtils.elapsedNanos(requestTimeNanos);
+        if (code != BKException.Code.OK) {
+            LOG.error(
+                    "Read of ledger entry failed: L{} E{}-E{}, Sent to {}, "
+                            + "Heard from {} : bitset = {}, Error = '{}'. First unread entry is ({}, rc = {})",
+                    lh.getId(), startEntryId, endEntryId, sentToHosts, heardFromHosts, heardFromHostsBitSet,
+                    BKException.getMessage(code), startEntryId, code);
+            clientCtx.getClientStats().getReadOpLogger().registerFailedEvent(latencyNanos, TimeUnit.NANOSECONDS);
+            // release the entries
+
+            request.close();
+            future.completeExceptionally(BKException.create(code));
+        } else {
+            clientCtx.getClientStats().getReadOpLogger().registerSuccessfulEvent(latencyNanos, TimeUnit.NANOSECONDS);
+            future.complete(LedgerEntriesImpl.create(request.entries));
+        }
+    }
+
+    BatchedReadOp parallelRead(boolean enabled) {
+        this.parallelRead = enabled;
+        return this;
+    }
+
+    @Override
+    public void readEntriesComplete(int rc, long ledgerId, long startEntryId, ByteBufList bufList, Object ctx) {
+        final ReadContext rctx = (ReadContext) ctx;
+        final BatchedLedgerEntryRequest entry = (BatchedLedgerEntryRequest) rctx.entry;
+
+        if (rc != BKException.Code.OK) {
```

Comment:
If the rc is not OK but the bufList is an empty list, it may lead to a direct memory leak.

Reply:
Ok. The empty bufList didn't occupy any memory. I will return null if the `rc != OK`